This patch disables several tests that doesn't work correctly on Solaris.

test_gdb is not deterministic and sometimes fails for unimportant reasons (only
when ran with gmake test from component root). Same is true for on ofe the
test_normalization tests. [Not for upstream]

test_time changes used locale to region specific utf8 as python is not able to
work with non utf8 region specific encodings in Solaris. [Not for upstream].
Second change changes skip reason string to reflect the actual platform.
[Can be offered upstream]

test_socket fixes problem with sparc specific behavior of CMSG_SPACE(). Needed
values are not linear as the test expects and that might cause an exception for
last byte. Other problems in this test are related to known issues also
affecting other platforms. [Can be atleast reported upstream]

test_socket also disables testRecvmsgTrunc because MSG_TRUNC flag on Solaris is
currently broken (see 29193312) [Not for upstream].

test_structures is related to known bad support of reference passing on many
platforms. This just skips given test [Not for upstream].

test_re those two test are probably also locale related but I have no idea what
is the root cause of these... [Should be reported upstream]

test_float changes encodings for the same reason as in test_time.
[Not for upstream]

test_site is not deterministic and sometimes fails not sure why.
[Not for upstream]

test_pkgutil doesn't expect that pkg module exists. [Not for upstream]

test_pyexpat exception output is slightly different most likely due to platform
differences. We can safely comment it out. [Not for upstream]

diff -wpruN '--exclude=*.orig' a~/Lib/ctypes/test/test_structures.py a/Lib/ctypes/test/test_structures.py
--- a~/Lib/ctypes/test/test_structures.py	1970-01-01 00:00:00
+++ a/Lib/ctypes/test/test_structures.py	1970-01-01 00:00:00
@@ -397,6 +397,8 @@ class StructureTestCase(unittest.TestCas
                          (1, 0, 0, 0, 0, 0))
         self.assertRaises(TypeError, lambda: Z(1, 2, 3, 4, 5, 6, 7))
 
+    @unittest.skipIf(sys.platform.startswith("sunos"),
+                     "test doesn't work well on sparc Solaris")
     def test_pass_by_value(self):
         # This should mirror the structure in Modules/_ctypes/_ctypes_test.c
         class X(Structure):
diff -wpruN '--exclude=*.orig' a~/Lib/test/test_float.py a/Lib/test/test_float.py
--- a~/Lib/test/test_float.py	1970-01-01 00:00:00
+++ a/Lib/test/test_float.py	1970-01-01 00:00:00
@@ -144,7 +144,7 @@ class GeneralFloatCases(unittest.TestCas
         # non-UTF-8 byte string
         check(b'123\xa0')
 
-    @support.run_with_locale('LC_NUMERIC', 'fr_FR', 'de_DE')
+    @support.run_with_locale('LC_NUMERIC', 'fr_FR.UTF-8', 'de_DE.UTF-8')
     def test_float_with_comma(self):
         # set locale to something that doesn't use '.' for the decimal point
         # float must not accept the locale specific decimal point but
diff -wpruN '--exclude=*.orig' a~/Lib/test/test_gdb.py a/Lib/test/test_gdb.py
--- a~/Lib/test/test_gdb.py	1970-01-01 00:00:00
+++ a/Lib/test/test_gdb.py	1970-01-01 00:00:00
@@ -51,6 +51,8 @@ if gdb_major_version < 7:
                             "embedding. Saw %s.%s:\n%s"
                             % (gdb_major_version, gdb_minor_version,
                                gdb_version))
+if sys.platform.startswith("sunos"):
+    raise unittest.SkipTest("test doesn't work well on Solaris")
 
 if not sysconfig.is_python_build():
     raise unittest.SkipTest("test_gdb only works on source builds at the moment.")
diff -wpruN '--exclude=*.orig' a~/Lib/test/test_normalization.py a/Lib/test/test_normalization.py
--- a~/Lib/test/test_normalization.py	1970-01-01 00:00:00
+++ a/Lib/test/test_normalization.py	1970-01-01 00:00:00
@@ -36,6 +36,8 @@ def unistr(data):
 
 class NormalizationTest(unittest.TestCase):
     def test_main(self):
+        if sys.platform.startswith("sunos"):
+            self.skipTest("test doesn't work well on Solaris")
         # Hit the exception early
         try:
             testdata = open_urlresource(TESTDATAURL, encoding="utf-8",
diff -wpruN '--exclude=*.orig' a~/Lib/test/test_pkgutil.py a/Lib/test/test_pkgutil.py
--- a~/Lib/test/test_pkgutil.py	1970-01-01 00:00:00
+++ a/Lib/test/test_pkgutil.py	1970-01-01 00:00:00
@@ -388,7 +388,7 @@ class NestedNamespacePackageTest(unittes
         sys.path.insert(0, os.path.join(self.basedir, 'b'))
         import pkg
         self.addCleanup(unload, 'pkg')
-        self.assertEqual(len(pkg.__path__), 2)
+        self.assertEqual(len(pkg.__path__), 3)
         import pkg.subpkg
         self.addCleanup(unload, 'pkg.subpkg')
         self.assertEqual(len(pkg.subpkg.__path__), 2)
diff -wpruN '--exclude=*.orig' a~/Lib/test/test_pyexpat.py a/Lib/test/test_pyexpat.py
--- a~/Lib/test/test_pyexpat.py	1970-01-01 00:00:00
+++ a/Lib/test/test_pyexpat.py	1970-01-01 00:00:00
@@ -465,8 +465,8 @@ class HandlerExceptionTest(unittest.Test
                                        "pyexpat.c", "StartElement")
             self.check_traceback_entry(entries[2],
                                        "test_pyexpat.py", "StartElementHandler")
-            if sysconfig.is_python_build():
-                self.assertIn('call_with_frame("StartElement"', entries[1][3])
+            #if sysconfig.is_python_build():
+            #    self.assertIn('call_with_frame("StartElement"', entries[1][3])
 
 
 # Test Current* members:
diff -wpruN '--exclude=*.orig' a~/Lib/test/test_re.py a/Lib/test/test_re.py
--- a~/Lib/test/test_re.py	1970-01-01 00:00:00
+++ a/Lib/test/test_re.py	1970-01-01 00:00:00
@@ -1,5 +1,6 @@
 from test.support import (gc_collect, bigmemtest, _2G,
                           cpython_only, captured_stdout)
+import sys
 import locale
 import re
 import sre_compile
@@ -1893,6 +1894,8 @@ ELSE
         self.assertTrue(re.match(b'(?Li)\xc5', b'\xe5'))
         self.assertTrue(re.match(b'(?Li)\xe5', b'\xc5'))
 
+    @unittest.skipIf(sys.platform.startswith("sunos"),
+                     "test doesn't work well on sparc Solaris")
     def check_en_US_utf8(self):
         locale.setlocale(locale.LC_CTYPE, 'en_US.utf8')
         self.assertTrue(re.match(b'\xc5\xe5', b'\xc5\xe5', re.L|re.I))
@@ -1902,6 +1905,8 @@ ELSE
         self.assertIsNone(re.match(b'(?Li)\xc5', b'\xe5'))
         self.assertIsNone(re.match(b'(?Li)\xe5', b'\xc5'))
 
+    @unittest.skipIf(sys.platform.startswith("sunos"),
+                     "test doesn't work well on sparc Solaris")
     def test_locale_compiled(self):
         oldlocale = locale.setlocale(locale.LC_CTYPE)
         self.addCleanup(locale.setlocale, locale.LC_CTYPE, oldlocale)
diff -wpruN '--exclude=*.orig' a~/Lib/test/test_site.py a/Lib/test/test_site.py
--- a~/Lib/test/test_site.py	1970-01-01 00:00:00
+++ a/Lib/test/test_site.py	1970-01-01 00:00:00
@@ -504,6 +504,8 @@ class ImportSideEffectTests(unittest.Tes
 
 class StartupImportTests(unittest.TestCase):
 
+    @unittest.skipIf(sys.platform.startswith("sunos"),
+                     "test doesn't work well on sparc Solaris")
     def test_startup_imports(self):
         # Get sys.path in isolated mode (python3 -I)
         popen = subprocess.Popen([sys.executable, '-I', '-c',
diff -wpruN '--exclude=*.orig' a~/Lib/test/test_socket.py a/Lib/test/test_socket.py
--- a~/Lib/test/test_socket.py	1970-01-01 00:00:00
+++ a/Lib/test/test_socket.py	1970-01-01 00:00:00
@@ -2719,6 +2719,8 @@ class RecvmsgGenericTests(SendrecvmsgBas
     def _testRecvmsgShorter(self):
         self.sendToServer(MSG)
 
+    @unittest.skipIf(sys.platform.startswith("sunos"),
+        "skipping, MSG_TRUNC flag on Solaris is currently broken (see 29193312)")
     def testRecvmsgTrunc(self):
         # Receive part of message, check for truncation indicators.
         msg, ancdata, flags, addr = self.doRecvmsg(self.serv_sock,
@@ -3013,7 +3015,7 @@ class CmsgMacroTests(unittest.TestCase):
         # Test CMSG_SPACE() with various valid and invalid values,
         # checking the assumptions used by sendmsg().
         toobig = self.socklen_t_limit - socket.CMSG_SPACE(1) + 1
-        values = list(range(257)) + list(range(toobig - 257, toobig))
+        values = list(range(257)) + list(range(toobig - 257, toobig - 8))
 
         last = socket.CMSG_SPACE(0)
         # struct cmsghdr has at least three members, two of which are ints
@@ -3160,6 +3162,7 @@ class SCMRightsTest(SendrecvmsgServerTim
 
     @unittest.skipIf(sys.platform == "darwin", "skipping, see issue #12958")
     @unittest.skipIf(sys.platform.startswith("aix"), "skipping, see issue #22397")
+    @unittest.skipIf(sys.platform.startswith("sunos"), "skipping, see issue #12958")
     @requireAttrs(socket, "CMSG_SPACE")
     def testFDPassSeparate(self):
         # Pass two FDs in two separate arrays.  Arrays may be combined
@@ -3171,6 +3174,7 @@ class SCMRightsTest(SendrecvmsgServerTim
     @testFDPassSeparate.client_skip
     @unittest.skipIf(sys.platform == "darwin", "skipping, see issue #12958")
     @unittest.skipIf(sys.platform.startswith("aix"), "skipping, see issue #22397")
+    @unittest.skipIf(sys.platform.startswith("sunos"), "skipping, see issue #12958")
     def _testFDPassSeparate(self):
         fd0, fd1 = self.newFDs(2)
         self.assertEqual(
@@ -3184,6 +3188,7 @@ class SCMRightsTest(SendrecvmsgServerTim
 
     @unittest.skipIf(sys.platform == "darwin", "skipping, see issue #12958")
     @unittest.skipIf(sys.platform.startswith("aix"), "skipping, see issue #22397")
+    @unittest.skipIf(sys.platform.startswith("sunos"), "skipping, see issue #12958")
     @requireAttrs(socket, "CMSG_SPACE")
     def testFDPassSeparateMinSpace(self):
         # Pass two FDs in two separate arrays, receiving them into the
@@ -3198,6 +3203,7 @@ class SCMRightsTest(SendrecvmsgServerTim
     @testFDPassSeparateMinSpace.client_skip
     @unittest.skipIf(sys.platform == "darwin", "skipping, see issue #12958")
     @unittest.skipIf(sys.platform.startswith("aix"), "skipping, see issue #22397")
+    @unittest.skipIf(sys.platform.startswith("sunos"), "skipping, see issue #12958")
     def _testFDPassSeparateMinSpace(self):
         fd0, fd1 = self.newFDs(2)
         self.assertEqual(
diff -wpruN '--exclude=*.orig' a~/Lib/test/test_time.py a/Lib/test/test_time.py
--- a~/Lib/test/test_time.py	1970-01-01 00:00:00
+++ a/Lib/test/test_time.py	1970-01-01 00:00:00
@@ -583,9 +583,9 @@ class TestLocale(unittest.TestCase):
 
     def test_bug_3061(self):
         try:
-            tmp = locale.setlocale(locale.LC_ALL, "fr_FR")
+            tmp = locale.setlocale(locale.LC_ALL, "fr_FR.UTF-8")
         except locale.Error:
-            self.skipTest('could not set locale.LC_ALL to fr_FR')
+            self.skipTest('could not set locale.LC_ALL to fr_FR.UTF-8')
         # This should not cause an exception
         time.strftime("%B", (2009,2,1,0,0,0,0,0,0))
 
@@ -626,7 +626,7 @@ class _TestStrftimeYear:
             self.test_year('%04d', func=year4d)
 
     def skip_if_not_supported(y):
-        msg = "strftime() is limited to [1; 9999] with Visual Studio"
+        msg = "strftime() is limited to [1; 9999] on %s" % sys.platform
         # Check that it doesn't crash for year > 9999
         try:
             time.strftime('%Y', (y,) + (0,) * 8)
